Objects/obmalloc.c

```diff
@@ -2663,9 +2683,71 @@ _PyDebugAllocatorStats(FILE *out,
     (void)printone(out, buf2, num_blocks * sizeof_block);
 }
 
+int _PyMem_init_obmalloc(PyInterpreterState *interp, _PyRuntimeState *runtime)
+{
+#ifdef WITH_PYMALLOC
+    /* Initialize obmalloc, but only for subinterpreters,
+       since the main interpreter is initialized statically. */
+    if (interp == &runtime->_main_interpreter
+            || (interp->feature_flags & Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
+        interp->obmalloc = &obmalloc_state_main;
+        interp->obmalloc->heap_allocated = false;
+    } else {
```

Comment:
```suggestion
    if (_Py_IsMainInterpreter(interp)
            || _PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        interp->obmalloc = &obmalloc_state_main;
        interp->obmalloc->heap_allocated = false;
    }
    else {
```

or keep the coupling as close as it was before:

```suggestion
    if (_Py_IsMainInterpreter(interp) {
        interp->obmalloc = &obmalloc_state_main;
        interp->obmalloc->heap_allocated = false;
    }
    else if (_PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        interp->obmalloc = _PyInterpreterState_Main()->obmalloc;
        assert(interp->obmalloc != NULL);
    }
    else {
```

or if you set `get_state()` back to the way it was (and return 0 for success):

```suggestion
    if (_Py_IsMainInterpreter(interp) {
        interp->obmalloc = &obmalloc_state_main;
        interp->obmalloc->heap_allocated = false;
    }
    else if (_PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        assert(interp->obmalloc == NULL);
        // It was already initialized for the main interpreter.
        return 0;
    }
    else {
```